ls1intum/Artemis
# Allow second correction for Exams

### Checklist

### Motivation and Context
In this PR we will enable the second correction feature for exams. 


### Description
If an exam has multiple correction rounds, tutors can now assess each submission multiple times. 
Once a first result exists, a second assessment is possible. Within the second assessment, the tutor will start with the preloaded first assessment. He can than edit the result just like before when working on an assessment. 

### Steps for Testing

The testing in this PR is very important. Please make sure to test it well and report all bugs or irregularities you notice. 
**1. Classic Handling**
1. Create an Exam with **_1 correction round_**. 
2. Everything should behave as usual, so you can lock assessments, assess them, submit them, and after that the student can complain about them.  

**2. New Handling**
1. Exam creation
1.1 Create an Exam with _**2** correction rounds_. 
1.2 Add a Programming, Text and Modeling Exercise.
2. Participate in the exam with different students.
3. Once the due date is over you can start assessing with tutors / instructors. 
4. Assessment - first correction round
4.1 start a new assessment in the first correction round. Go back to the assessment dashboard and check if the assessment is locked
4.2 continue / start a new assessment, add feedback and save. Leave and then return to the assessment view again, and check if the save worked.
4.3 Submit a assessment
4.4 Use Override and Next Assessment
5. Assessment - second correction round
5.1 After a submission is assessed, and has a result in the first correction round you can assess in the second correction round. (It is still possible that a tutor gets the same submission in the second correction round that he has already evaluated in the first one. This will be changed in the future.) 
5.2 When starting a 2. correction, you should receive your assessment with the results of the first correction preloaded. 
5.3 now repeat all steps from 4 in the second correction round, editing the result should work as usual.
6. After the assessment is finished use the student accounts to view the result. Only the result of the second correction should be displayed.
7. Create complaints for each submission, process those complaints with tutors
8. The students should then see the result, with the complaint status

### Test Coverage
exercise-assessment-dashboard.component.ts    83.9%
ProgrammingSubmissionResource: 84%
TextSubmissionResource: 92 %
FileUploadSubmissionResource: 83% 
ModelingSubmissionResource: 97% 
SubmissionResource: 90%

### Screenshots


![Bildschirmfoto 2021-01-10 um 23 24 51](https://user-images.githubusercontent.com/42205689/104137166-2e7e5680-539b-11eb-8a54-67085c9e92e6.png)
